src/graph/bellman_ford.md

```diff
@@ -208,6 +208,53 @@ The above implementation looks for a negative cycle reachable from some starting
 
 For more on this topic — see separate article, [Finding a negative cycle in the graph](./graph/finding-negative-cycle-in-graph.html).
 
+## Shortest Path Faster Algorithm (SPFA)
+
+SPFA is a improvement of the Bellman-Ford algorithm which takes advantage of the fact that not all attempts at relaxation will work. The main idea is to create a queue containing only the vertices that were relaxed but that still could not relax their neighbors and whenever you can relax some neighbor you should put him in the queue. This algorithm can also be used to detect negative cycles as the Bellman-Ford.
+
+The worst case of this algorithm is equal to the $O(nm)$ of the Bellman-Ford but in practice it works in much better speed and there are even [people who say it works in $O(m)$ on average](https://en.wikipedia.org/wiki/Shortest_Path_Faster_Algorithm#Average-case_performance).
```

Comment:
*warning*: it is quite easy nowadays to make testcases for SPFA implementations to run in O(nm), I think it might be better to warn users of this.

Reply:
Thanks, I did not know this, I'll add the warning. Can you tell me some problems that use these types of testcases ?

Reply:
@wilwxk (discussion in chinese) https://www.zhihu.com/question/292283275/answer/484694411
(onlinejudge) https://duck.ac/problem/noi18a